network/transport/v2/handlers.go

```diff
@@ -47,32 +48,66 @@ func (p protocol) Handle(peer transport.Peer, raw interface{}) error {
 	return nil
 }
 
-func (p protocol) handle(peer transport.Peer, envelope *Envelope) error {
-	switch msg := envelope.Message.(type) {
+type handleFunc func(peer transport.Peer, envelope *Envelope) error
+
+func (p *protocol) handleASynchWithLock(peer transport.Peer, envelope *Envelope, f handleFunc) error {
+	go func() {
+		if !p.mLock.TryRLock() {
+			log.Logger().Debugf("Sink active, ignoring %T (peer=%s)", envelope.Message, peer)
+			return
+		}
+		defer p.mLock.RUnlock()
+		if err := f(peer, envelope); err != nil {
+			log.Logger().Errorf("error handling %T (peer=%s): %s", envelope.Message, peer, err)
+		}
+	}()
+	return nil
+}
+
+func handleASync(peer transport.Peer, envelope *Envelope, f handleFunc) error {
+	go func() {
+		if err := f(peer, envelope); err != nil {
+			log.Logger().Errorf("error handling %T (peer=%s): %s", envelope.Message, peer, err)
+		}
+	}()
+	return nil
+}
+
+func (p *protocol) handle(peer transport.Peer, envelope *Envelope) error {
+	switch envelope.Message.(type) {
 	case *Envelope_Gossip:
-		return p.handleGossip(peer, msg.Gossip)
+		return p.handleASynchWithLock(peer, envelope, p.handleGossip)
 	case *Envelope_Hello:
 		log.Logger().Infof("%T: %s said hello", p, peer)
 		return nil
 	case *Envelope_TransactionList:
-		return p.handleTransactionList(peer, msg)
+		// in order handling of transactionLists
+		// the transactionList handler locks after parsing and before DB access.
+		p.listHandler.ch <- peerEnvelope{
+			envelope: envelope,
+			peer:     peer,
+		}
+		return nil
 	case *Envelope_TransactionListQuery:
-		return p.handleTransactionListQuery(peer, msg.TransactionListQuery)
+		return p.handleASynchWithLock(peer, envelope, p.handleTransactionListQuery)
 	case *Envelope_TransactionPayloadQuery:
-		return p.handleTransactionPayloadQuery(peer, msg.TransactionPayloadQuery)
+		return p.handleASynchWithLock(peer, envelope, p.handleTransactionPayloadQuery)
 	case *Envelope_TransactionPayload:
-		return p.handleTransactionPayload(msg.TransactionPayload)
+		return p.handleASynchWithLock(peer, envelope, p.handleTransactionPayload)
 	case *Envelope_TransactionRangeQuery:
-		return p.handleTransactionRangeQuery(peer, msg)
+		return p.handleASynchWithLock(peer, envelope, p.handleTransactionRangeQuery)
 	case *Envelope_State:
-		return p.handleState(peer, msg.State)
+		// no DB interaction so no lock required
+		return handleASync(peer, envelope, p.handleState)
 	case *Envelope_TransactionSet:
-		return p.handleTransactionSet(peer, msg)
+		// no DB interaction so no lock required
+		return handleASync(peer, envelope, p.handleTransactionSet)
```

Comment:
`p.state.XOR()` and `p.state.IBLT()` currently use a cursor to get the highest LC from the DB when the requested clock value is `math.MaxUint32`. State should probably keep track of the highest LC to reduce DB interactions.

Reply:
Didn't know that, but yes that would be a very good optimization. Until then these will have to lock as well.

Reply:
added #1055